### PR TITLE
Add RBAC data model, migration, and store CRUD

### PR DESCRIPTION
**Pull request by PattaraS**
## 🥞 Stacked PR
Use this [link](https://github.com/mlflow/mlflow/pull/22721/files) to review incremental changes.
- [**stack/rbac/data-model**](https://github.com/mlflow/mlflow/pull/22721) [[Files changed](https://github.com/mlflow/mlflow/pull/22721/files)]
  - [stack/rbac/rest-api](https://github.com/mlflow/mlflow/pull/22722) [[Files changed](https://github.com/mlflow/mlflow/pull/22722/files/225bd25ec2fb906f707fb7f2c5d3ee56c8d87bad..9eeedfdb4b715e8256a94318d4005534dc346f7d)]

---------
### Related Issues/PRs

Relates to user management & access control feature (RBAC milestone)

### What changes are proposed in this pull request?

This is **PR 1/2** in the RBAC stack. It adds the data layer for Role-Based Access Control:

- **New entity classes** (`Role`, `RolePermission`, `UserRoleAssignment`) in `entities.py`
- **New SQLAlchemy models** (`SqlRole`, `SqlRolePermission`, `SqlUserRoleAssignment`) in `db/models.py`
- **Alembic migration** creating 3 new tables: `roles`, `role_permissions`, `user_role_assignments`
- **Permission helpers** in `permissions.py`: `PERMISSION_PRIORITY`, `max_permission()`, `VALID_RESOURCE_TYPES`, `_validate_resource_type()`
- **Full CRUD** in `SqlAlchemyStore`: role/role_permission/user_role_assignment management
- **Role-based permission resolution**: `get_role_permission_for_resource()` queries the user's roles in a workspace, checks for workspace-wide grants, matches specific or wildcard (`*`) resource patterns, and returns the highest permission via union

Key design decisions:
- Roles are **workspace-scoped** — a role belongs to exactly one workspace
- **Wildcard resource patterns**: `resource_pattern = "*"` grants access to all resources of that type in the workspace
- **Workspace admin is expressed as a role permission**: `(resource_type="workspace", resource_pattern="*", permission=MANAGE)` grants workspace-wide MANAGE (including role and user-assignment management within the workspace). Lower permission levels on `resource_type="workspace"` apply to every resource type in the workspace
- **Permission union**: multiple roles' permissions are combined, taking the highest
- **Additive**: the direct per-resource permission tables remain the primary store for existing grants; roles layer on top. A follow-up migration will consolidate the per-resource tables into `role_permissions` via synthetic per-user roles

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests

51 unit tests in `tests/server/auth/test_sqlalchemy_store_rbac.py` covering:
- Role CRUD (create, get, list, update, delete, cascade)
- Role permission CRUD (add, remove, list, update, wildcards, validation)
- User-role assignment CRUD (assign, unassign, list by user/workspace/role, nonexistent user)
- Permission resolution (specific match, wildcard, union of multiple roles, workspace-scope applying across types, cross-workspace isolation)
- `is_workspace_admin` check (MANAGE-only, negative for non-MANAGE)

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] No.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

**Comment by PattaraS** [comment]
<details>
<summary>Range-diff: master (db15a4f -&gt; 270f386)</summary>

> <details>
> <summary>mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py</summary>
>
> ````diff
> @@ -27,7 +27,7 @@
>  +        sa.Column("name", sa.String(length=255), nullable=False),
>  +        sa.Column("workspace", sa.String(length=63), nullable=False),
>  +        sa.Column("description", sa.String(length=1024), nullable=True),
> -+        sa.Column("is_workspace_admin", sa.Boolean(), server_default=sa.text("0")),
> ++        sa.Column("is_workspace_admin", sa.Boolean(), server_default=sa.false()),
>  +        sa.UniqueConstraint("workspace", "name", name="unique_workspace_role_name"),
>  +    )
>  +    op.create_index("idx_roles_workspace", "roles", ["workspace"], unique=False)
> ````
>
> </details>

> <details>
> <summary>mlflow/server/auth/sqlalchemy_store.py</summary>
>
> ````diff
> @@ -1,6 +1,12 @@
>  diff --git a/mlflow/server/auth/sqlalchemy_store.py b/mlflow/server/auth/sqlalchemy_store.py
>  --- a/mlflow/server/auth/sqlalchemy_store.py
>  +++ b/mlflow/server/auth/sqlalchemy_store.py
> + from sqlalchemy.exc import IntegrityError, MultipleResultsFound, NoResultFound
> +-from sqlalchemy.orm import sessionmaker
> ++from sqlalchemy.orm import selectinload, sessionmaker
> + from werkzeug.security import check_password_hash, generate_password_hash
> + 
> + from mlflow.environment_variables import MLFLOW_ENABLE_WORKSPACES
>       SqlGatewayModelDefinitionPermission,
>       SqlGatewaySecretPermission,
>       SqlRegisteredModelPermission,
> @@ -144,9 +150,9 @@
>  +
>  +    def delete_roles_for_workspace(self, workspace_name: str) -> None:
>  +        with self.ManagedSessionMaker() as session:
> -+            session.query(SqlRole).filter(SqlRole.workspace == workspace_name).delete(
> -+                synchronize_session=False
> -+            )
> ++            roles = session.query(SqlRole).filter(SqlRole.workspace == workspace_name).all()
> ++            for role in roles:
> ++                session.delete(role)
>  +
>  +    # ---- RolePermission CRUD ----
>  +
> @@ -224,22 +230,37 @@
>  +
>  +    def assign_role_to_user(self, user_id: int, role_id: int) -> UserRoleAssignment:
>  +        with self.ManagedSessionMaker() as session:
> ++            # Validate both user and role exist before attempting assignment
> ++            user = session.get(SqlUser, user_id)
> ++            if user is None:
> ++                raise MlflowException(
> ++                    f"User with id={user_id} not found",
> ++                    RESOURCE_DOES_NOT_EXIST,
> ++                )
>  +            self._get_role(session, role_id)
> -+            try:
> -+                assignment = SqlUserRoleAssignment(user_id=user_id, role_id=role_id)
> -+                session.add(assignment)
> -+                session.flush()
> -+                return UserRoleAssignment(
> -+                    id_=assignment.id,
> -+                    user_id=assignment.user_id,
> -+                    role_id=assignment.role_id,
> ++            # Check for duplicate assignment before insert
> ++            existing = (
> ++                session
> ++                .query(SqlUserRoleAssignment)
> ++                .filter(
> ++                    SqlUserRoleAssignment.user_id == user_id,
> ++                    SqlUserRoleAssignment.role_id == role_id,
>  +                )
> -+            except IntegrityError as e:
> ++                .first()
> ++            )
> ++            if existing is not None:
>  +                raise MlflowException(
> -+                    f"User role assignment (user_id={user_id}, role_id={role_id}) "
> -+                    f"already exists. Error: {e}",
> ++                    f"User role assignment (user_id={user_id}, role_id={role_id}) already exists",
>  +                    RESOURCE_ALREADY_EXISTS,
> -+                ) from e
> ++                )
> ++            assignment = SqlUserRoleAssignment(user_id=user_id, role_id=role_id)
> ++            session.add(assignment)
> ++            session.flush()
> ++            return UserRoleAssignment(
> ++                id_=assignment.id,
> ++                user_id=assignment.user_id,
> ++                role_id=assignment.role_id,
> ++            )
>  +
>  +    def unassign_role_from_user(self, user_id: int, role_id: int) -> None:
>  +        with self.ManagedSessionMaker() as session:
> @@ -311,6 +332,7 @@
>  +            roles = (
>  +                session
>  +                .query(SqlRole)
> ++                .options(selectinload(SqlRole.permissions))
>  +                .join(SqlUserRoleAssignment, SqlRole.id == SqlUserRoleAssignment.role_id)
>  +                .filter(
>  +                    SqlUserRoleAssignment.user_id == user_id,
> ````
>
> </details>

> <details>
> <summary>tests/server/auth/test_sqlalchemy_store_rbac.py</summary>
>
> ````diff
> @@ -151,6 +151,17 @@
>  +    assert len(store.list_roles("ws2")) == 1
>  +
>  +
> ++def test_delete_roles_for_workspace_cascades(store, user):
> ++    role = store.create_role(name="r1", workspace="ws1")
> ++    store.add_role_permission(role.id, "experiment", "*", "READ")
> ++    store.assign_role_to_user(user.id, role.id)
> ++
> ++    store.delete_roles_for_workspace("ws1")
> ++
> ++    assert store.list_roles("ws1") == []
> ++    assert store.list_user_roles(user.id) == []
> ++
> ++
>  +# ---- RolePermission CRUD ----
>  +
>  +
> @@ -246,6 +257,12 @@
>  +    assert assignment.role_id == role.id
>  +
>  +
> ++def test_assign_role_nonexistent_user(store):
> ++    role = store.create_role(name="viewer", workspace="ws1")
> ++    with pytest.raises(MlflowException, match="not found"):
> ++        store.assign_role_to_user(99999, role.id)
> ++
> ++
>  +def test_assign_role_duplicate(store, user):
>  +    role = store.create_role(name="viewer", workspace="ws1")
>  +    store.assign_role_to_user(user.id, role.id)
> ````
>
> </details>


<sub>Reproduce locally: `git range-diff 1dfd617..db15a4f 1dfd617..270f386` | Disable: `git config gitstack.push-range-diff false`</sub>
</details>

**Comment by PattaraS** [comment]
<details>
<summary>Range-diff: master (270f386 -&gt; 29848bf)</summary>

> <details>
> <summary>mlflow/server/auth/db/models.py</summary>
>
> ````diff
> @@ -8,7 +8,10 @@
>  +    RolePermission,
>       ScorerPermission,
>       User,
> ++    UserRoleAssignment,
>       WorkspacePermission,
> + )
> + from mlflow.utils.workspace_utils import DEFAULT_WORKSPACE_NAME
>               user_id=self.user_id,
>               permission=self.permission,
>           )
> @@ -77,4 +80,11 @@
>  +        UniqueConstraint("user_id", "role_id", name="unique_user_role"),
>  +        Index("idx_user_role_assignments_user_id", "user_id"),
>  +        Index("idx_user_role_assignments_role_id", "role_id"),
> -+    )
> \ No newline at end of file
> ++    )
> ++
> ++    def to_mlflow_entity(self):
> ++        return UserRoleAssignment(
> ++            id_=self.id,
> ++            user_id=self.user_id,
> ++            role_id=self.role_id,
> ++        )
> \ No newline at end of file
> ````
>
> </details>

> <details>
> <summary>mlflow/server/auth/sqlalchemy_store.py</summary>
>
> ````diff
> @@ -79,6 +79,11 @@
>  +                f"Role with id={role_id} not found",
>  +                RESOURCE_DOES_NOT_EXIST,
>  +            )
> ++        except MultipleResultsFound:
> ++            raise MlflowException(
> ++                f"Found multiple roles with id={role_id}",
> ++                INVALID_STATE,
> ++            )
>  +
>  +    @staticmethod
>  +    def _get_role_by_name(session, workspace: str, name: str) -> SqlRole:
> @@ -94,6 +99,11 @@
>  +                f"Role with name={name} in workspace={workspace} not found",
>  +                RESOURCE_DOES_NOT_EXIST,
>  +            )
> ++        except MultipleResultsFound:
> ++            raise MlflowException(
> ++                f"Found multiple roles with name={name} in workspace={workspace}",
> ++                INVALID_STATE,
> ++            )
>  +
>  +    def get_role(self, role_id: int) -> Role:
>  +        with self.ManagedSessionMaker() as session:
> @@ -256,11 +266,7 @@
>  +            assignment = SqlUserRoleAssignment(user_id=user_id, role_id=role_id)
>  +            session.add(assignment)
>  +            session.flush()
> -+            return UserRoleAssignment(
> -+                id_=assignment.id,
> -+                user_id=assignment.user_id,
> -+                role_id=assignment.role_id,
> -+            )
> ++            return assignment.to_mlflow_entity()
>  +
>  +    def unassign_role_from_user(self, user_id: int, role_id: int) -> None:
>  +        with self.ManagedSessionMaker() as session:
> @@ -283,16 +289,13 @@
>  +
>  +    def list_user_roles(self, user_id: int) -> list[Role]:
>  +        with self.ManagedSessionMaker() as session:
> -+            assignments = (
> ++            roles = (
>  +                session
> -+                .query(SqlUserRoleAssignment)
> ++                .query(SqlRole)
> ++                .join(SqlUserRoleAssignment, SqlRole.id == SqlUserRoleAssignment.role_id)
>  +                .filter(SqlUserRoleAssignment.user_id == user_id)
>  +                .all()
>  +            )
> -+            role_ids = [a.role_id for a in assignments]
> -+            if not role_ids:
> -+                return []
> -+            roles = session.query(SqlRole).filter(SqlRole.id.in_(role_ids)).all()
>  +            return [r.to_mlflow_entity() for r in roles]
>  +
>  +    def list_user_roles_for_workspace(self, user_id: int, workspace: str) -> list[Role]:
> @@ -318,10 +321,7 @@
>  +                .filter(SqlUserRoleAssignment.role_id == role_id)
>  +                .all()
>  +            )
> -+            return [
> -+                UserRoleAssignment(id_=a.id, user_id=a.user_id, role_id=a.role_id)
> -+                for a in assignments
> -+            ]
> ++            return [a.to_mlflow_entity() for a in assignments]
>  +
>  +    # ---- Role-based permission resolution ----
>  +
> ````
>
> </details>


<sub>Reproduce locally: `git range-diff 1dfd617..270f386 1dfd617..29848bf` | Disable: `git config gitstack.push-range-diff false`</sub>
</details>

**Comment by PattaraS** [comment]
/review

---

✅ [Review](https://github.com/mlflow/mlflow/actions/runs/24706697062?pr=22721) completed
<details><summary>Review Output</summary>

PR reviewed and approved. Here's the summary:

**Overall**: Clean RBAC data layer with well-structured entities, models, migration, and store CRUD. 48 tests provide solid coverage.

**Comment added**: `list_user_roles` should use a single join (like `list_user_roles_for_workspace`) instead of two queries.

**Noted in approval**: Several earlier copilot comments are already fixed in the current revision. Two remaining items worth addressing before merge:
1. `assign_role_to_user` flush needs `IntegrityError` handling for concurrent duplicates
2. `list_user_roles` / `list_user_roles_for_workspace` should validate `user_id` exists for consistency with other store methods

</details>

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        sa.Column("id", sa.Integer(), primary_key=True),
+        sa.Column("name", sa.String(length=255), nullable=False),
+        sa.Column("workspace", sa.String(length=63), nullable=False),
+        sa.Column("description", sa.String(length=1024), nullable=True),
+        sa.Column("is_workspace_admin", sa.Boolean(), server_default=sa.false()),
+        sa.UniqueConstraint("workspace", "name", name="unique_workspace_role_name"),
+    )
+    op.create_index("idx_roles_workspace", "roles", ["workspace"], unique=False)
+
+    op.create_table(
```
Comment (TomeHirata) [inline mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:31]: I like this structure. Do we want to migrate the current per-resource tables to this table too?

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def upgrade() -> None:
+    op.create_table(
+        "roles",
+        sa.Column("id", sa.Integer(), primary_key=True),
+        sa.Column("name", sa.String(length=255), nullable=False),
+        sa.Column("workspace", sa.String(length=63), nullable=False),
+        sa.Column("description", sa.String(length=1024), nullable=True),
+        sa.Column("is_workspace_admin", sa.Boolean(), server_default=sa.false()),
```
Comment (TomeHirata) [inline mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:26]: Do we want to represent this by role_permissions where permission=admin/manage?

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+down_revision = "2ed73881770d"
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    op.create_table(
+        "roles",
+        sa.Column("id", sa.Integer(), primary_key=True),
+        sa.Column("name", sa.String(length=255), nullable=False),
```
Comment (TomeHirata) [inline mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:23]: Shouldn't role name be non-null?

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        sa.Column("workspace", sa.String(length=63), nullable=False),
+        sa.Column("description", sa.String(length=1024), nullable=True),
```
Comment (TomeHirata) [inline mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:25]: How do we use description?

**Review by TomeHirata — APPROVED** [review]
LGTM once comments are addressed/answered!

**Comment by PattaraS** [comment]
<details>
<summary>Range-diff: master (29848bf -&gt; 2e0ac06)</summary>

> <details>
> <summary>tests/server/auth/db/test_cli.py</summary>
>
> ````diff
> @@ -0,0 +1,18 @@
> +diff --git a/tests/server/auth/db/test_cli.py b/tests/server/auth/db/test_cli.py
> +--- a/tests/server/auth/db/test_cli.py
> ++++ b/tests/server/auth/db/test_cli.py
> +         ("gateway_endpoint_permissions",),
> +         ("gateway_model_definition_permissions",),
> +         ("workspace_permissions",),
> ++        ("roles",),
> ++        ("role_permissions",),
> ++        ("user_role_assignments",),
> +     ])
> + 
> + 
> +     assert "scorer_permissions" in tables
> +     assert "registered_model_permissions" in tables
> +     assert "workspace_permissions" in tables
> +-    assert version[0] == "2ed73881770d"
> ++    assert version[0] == "c3d4e5f6a7b8"
> +     assert user == ("testuser", 1)
> \ No newline at end of file
> ````
>
> </details>


<sub>Reproduce locally: `git range-diff 1dfd617..29848bf 1dfd617..2e0ac06` | Disable: `git config gitstack.push-range-diff false`</sub>
</details>

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py (hunk 3 above)
Comment (PattaraS) [reply mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:23]: The `name` column already has `nullable=False` on that line (`sa.Column("name", sa.String(length=255), nullable=False)`). Were you looking at a different version?

🤖 *Claude response*

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py (hunk 4 above)
Comment (PattaraS) [reply mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:25]: Free-form text to describe the role's purpose (e.g., "Data scientist: read+edit on ML experiments in this workspace"). Optional, shown in the admin UI for operator context.

🤖 *Claude response*

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py (hunk 2 above)
Comment (PattaraS) [reply mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:26]: Considered this but kept the boolean for two reasons:

1. WP admin semantics go beyond resource-level permissions — a WP admin can also create/update roles and assign users to roles, which aren't resource-scoped actions. Expressing those with `role_permissions` would require a synthetic resource type.
2. A boolean lets us short-circuit in permission resolution (return MANAGE immediately) without loading and scanning all `role_permissions` rows for the role.

Happy to revisit if you feel strongly.

🤖 *Claude response*

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py (hunk 1 above)
Comment (PattaraS) [reply mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:31]: Agreed with the direction. The per-resource tables are really an ad-hoc way to represent "each user's personal permission set," which doesn't scale to new resource types. Proposed follow-up PR:

- Create a synthetic internal role per user, e.g. `__user_<user_id>__` (hidden from admin UI listings)
- Migrate existing per-resource grants into `role_permissions` rows on that role
- Auto-assign every user to their own internal role on creation
- Update permission resolution to no longer read from the per-resource tables
- Remove the per-resource tables in a subsequent cleanup PR (after confirming no callers depend on them)

I'd prefer to do this as a separate PR to keep this one focused on introducing the role infrastructure. Sound good?

🤖 *Claude response*

**Comment by PattaraS** [comment]
<details>
<summary>Range-diff: master (2e0ac06 -&gt; 21c0428)</summary>

> <details>
> <summary>docs/api_reference/api_inventory.txt</summary>
>
> ````diff
> @@ -0,0 +1,24 @@
> +diff --git a/docs/api_reference/api_inventory.txt b/docs/api_reference/api_inventory.txt
> +--- a/docs/api_reference/api_inventory.txt
> ++++ b/docs/api_reference/api_inventory.txt
> + mlflow.server.auth.entities.RegisteredModelPermission
> + mlflow.server.auth.entities.RegisteredModelPermission.from_json
> + mlflow.server.auth.entities.RegisteredModelPermission.to_json
> ++mlflow.server.auth.entities.Role
> ++mlflow.server.auth.entities.Role.from_json
> ++mlflow.server.auth.entities.Role.to_json
> ++mlflow.server.auth.entities.RolePermission
> ++mlflow.server.auth.entities.RolePermission.from_json
> ++mlflow.server.auth.entities.RolePermission.to_json
> + mlflow.server.auth.entities.ScorerPermission
> + mlflow.server.auth.entities.ScorerPermission.from_json
> + mlflow.server.auth.entities.ScorerPermission.to_json
> + mlflow.server.auth.entities.User
> + mlflow.server.auth.entities.User.from_json
> + mlflow.server.auth.entities.User.to_json
> ++mlflow.server.auth.entities.UserRoleAssignment
> ++mlflow.server.auth.entities.UserRoleAssignment.from_json
> ++mlflow.server.auth.entities.UserRoleAssignment.to_json
> + mlflow.server.auth.entities.WorkspacePermission
> + mlflow.server.auth.entities.WorkspacePermission.from_json
> + mlflow.server.auth.entities.WorkspacePermission.to_json
> \ No newline at end of file
> ````
>
> </details>


<sub>Reproduce locally: `git range-diff 1dfd617..2e0ac06 1dfd617..21c0428` | Disable: `git config gitstack.push-range-diff false`</sub>
</details>

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py (hunk 1 above)
Comment (TomeHirata) [reply mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:30]: Yeah definitely not in this PR's scope

File: mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py (hunk 3 above)
Comment (TomeHirata) [reply mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py:23]: my bad

**Comment by PattaraS** [comment]
<details>
<summary>Range-diff: master (21c0428 -&gt; 225bd25)</summary>

> <details>
> <summary>mlflow/server/auth/db/migrations/versions/c3d4e5f6a7b8_add_rbac_tables.py</summary>
>
> ````diff
> @@ -27,7 +27,6 @@
>  +        sa.Column("name", sa.String(length=255), nullable=False),
>  +        sa.Column("workspace", sa.String(length=63), nullable=False),
>  +        sa.Column("description", sa.String(length=1024), nullable=True),
> -+        sa.Column("is_workspace_admin", sa.Boolean(), server_default=sa.false()),
>  +        sa.UniqueConstraint("workspace", "name", name="unique_workspace_role_name"),
>  +    )
>  +    op.create_index("idx_roles_workspace", "roles", ["workspace"], unique=False)
> ````
>
> </details>

> <details>
> <summary>mlflow/server/auth/db/models.py</summary>
>
> ````diff
> @@ -24,7 +24,6 @@
>  +    name = Column(String(255), nullable=False)
>  +    workspace = Column(String(63), nullable=False)
>  +    description = Column(String(1024), nullable=True)
> -+    is_workspace_admin = Column(Boolean, default=False)
>  +    permissions = relationship("SqlRolePermission", backref="role", cascade="all, delete-orphan")
>  +    user_assignments = relationship(
>  +        "SqlUserRoleAssignment", backref="role", cascade="all, delete-orphan"
> @@ -40,7 +39,6 @@
>  +            name=self.name,
>  +            workspace=self.workspace,
>  +            description=self.description,
> -+            is_workspace_admin=self.is_workspace_admin,
>  +            permissions=[p.to_mlflow_entity() for p in self.permissions],
>  +        )
>  +
> ````
>
> </details>

> <details>
> <summary>mlflow/server/auth/entities.py</summary>
>
> ````diff
> @@ -18,14 +18,12 @@
>  +        name,
>  +        workspace,
>  +        description=None,
> -+        is_workspace_admin=False,
>  +        permissions=None,
>  +    ):
>  +        self._id = id_
>  +        self._name = name
>  +        self._workspace = workspace
>  +        self._description = description
> -+        self._is_workspace_admin = is_workspace_admin
>  +        self._permissions = permissions or []
>  +
>  +    @property
> @@ -51,10 +49,6 @@
>  +    @description.setter
>  +    def description(self, description):
>  +        self._description = description
> -+
> -+    @property
> -+    def is_workspace_admin(self):
> -+        return self._is_workspace_admin
>  +
>  +    @property
>  +    def permissions(self):
> @@ -66,7 +60,6 @@
>  +            "name": self.name,
>  +            "workspace": self.workspace,
>  +            "description": self.description,
> -+            "is_workspace_admin": self.is_workspace_admin,
>  +            "permissions": [p.to_json() for p in self.permissions],
>  +        }
>  +
> @@ -77,7 +70,6 @@
>  +            name=dictionary["name"],
>  +            workspace=dictionary.get("workspace", DEFAULT_WORKSPACE_NAME),
>  +            description=dictionary.get("description"),
> -+            is_workspace_admin=dictionary.get("is_workspace_admin", False),
>  +            permissions=[RolePermission.from_json(p) for p in dictionary.get("permissions", [])],
>  +        )
>  +
> ````
>
> </details>

> <details>
> <summary>mlflow/server/auth/permissions.py</summary>
>
> ````diff
> @@ -19,6 +19,10 @@
>  +    "gateway_secret",
>  +    "gateway_endpoint",
>  +    "gateway_model_definition",
> ++    # Special: workspace-wide permissions that apply to all resources in the role's workspace.
> ++    # resource_pattern must be "*". permission=MANAGE additionally grants role/user
> ++    # management capabilities within the workspace.
> ++    "workspace",
>  +})
>  +
>  +
> ````
>
> </details>

> <details>
> <summary>mlflow/server/auth/sqlalchemy_store.py</summary>
>
> ````diff
> @@ -51,7 +51,6 @@
>  +        name: str,
>  +        workspace: str,
>  +        description: str | None = None,
> -+        is_workspace_admin: bool = False,
>  +    ) -> Role:
>  +        with self.ManagedSessionMaker() as session:
>  +            try:
> @@ -59,7 +58,6 @@
>  +                    name=name,
>  +                    workspace=workspace,
>  +                    description=description,
> -+                    is_workspace_admin=is_workspace_admin,
>  +                )
>  +                session.add(role)
>  +                session.flush()
> @@ -345,10 +343,16 @@
>  +
>  +            best_permission_name: str | None = None
>  +            for role in roles:
> -+                if role.is_workspace_admin:
> -+                    return MANAGE
> -+
>  +                for rp in role.permissions:
> ++                    # Workspace-wide permission — applies to every resource type.
> ++                    if rp.resource_type == "workspace" and rp.resource_pattern == "*":
> ++                        best_permission_name = (
> ++                            max_permission(best_permission_name, rp.permission)
> ++                            if best_permission_name is not None
> ++                            else rp.permission
> ++                        )
> ++                        continue
> ++                    # Resource-type-specific permission.
>  +                    if rp.resource_type != resource_type:
>  +                        continue
>  +                    if rp.resource_pattern in ("*", resource_id):
> @@ -360,4 +364,26 @@
>  +
>  +            if best_permission_name is None:
>  +                return None
> -+            return get_permission(best_permission_name)
> \ No newline at end of file
> ++            return get_permission(best_permission_name)
> ++
> ++    def is_workspace_admin(self, user_id: int, workspace: str) -> bool:
> ++        """
> ++        True if the user has a role in the given workspace with
> ++        (resource_type='workspace', resource_pattern='*', permission=MANAGE).
> ++        """
> ++        with self.ManagedSessionMaker() as session:
> ++            return (
> ++                session
> ++                .query(SqlRolePermission)
> ++                .join(SqlRole, SqlRole.id == SqlRolePermission.role_id)
> ++                .join(SqlUserRoleAssignment, SqlRole.id == SqlUserRoleAssignment.role_id)
> ++                .filter(
> ++                    SqlUserRoleAssignment.user_id == user_id,
> ++                    SqlRole.workspace == workspace,
> ++                    SqlRolePermission.resource_type == "workspace",
> ++                    SqlRolePermission.resource_pattern == "*",
> ++                    SqlRolePermission.permission == MANAGE.name,
> ++                )
> ++                .first()
> ++                is not None
> ++            )
> \ No newline at end of file
> ````
>
> </details>

> <details>
> <summary>tests/server/auth/test_sqlalchemy_store_rbac.py</summary>
>
> ````diff
> @@ -40,7 +40,6 @@
>  +    assert role.name == "viewer"
>  +    assert role.workspace == "ws1"
>  +    assert role.description == "Read-only access"
> -+    assert role.is_workspace_admin is False
>  +    assert role.permissions == []
>  +
>  +
> @@ -56,11 +55,6 @@
>  +    assert r1.id != r2.id
>  +
>  +
> -+def test_create_workspace_admin_role(store):
> -+    role = store.create_role(name="ws-admin", workspace="ws1", is_workspace_admin=True)
> -+    assert role.is_workspace_admin is True
> -+
> -+
>  +def test_get_role(store):
>  +    created = store.create_role(name="editor", workspace="ws1")
>  +    fetched = store.get_role(created.id)
> @@ -382,12 +376,58 @@
>  +
>  +
>  +def test_get_role_permission_workspace_admin(store, user):
> -+    role = store.create_role(name="ws-admin", workspace="ws1", is_workspace_admin=True)
> ++    role = store.create_role(name="ws-admin", workspace="ws1")
> ++    store.add_role_permission(role.id, "workspace", "*", "MANAGE")
>  +    store.assign_role_to_user(user.id, role.id)
>  +
> ++    # Workspace-wide MANAGE applies to any resource type.
>  +    result = store.get_role_permission_for_resource(user.id, "experiment", "any-id", "ws1")
>  +    assert result == MANAGE
>  +
> ++    result = store.get_role_permission_for_resource(user.id, "registered_model", "m1", "ws1")
> ++    assert result == MANAGE
> ++
> ++
> ++def test_workspace_permission_applies_across_resource_types(store, user):
> ++    role = store.create_role(name="reader", workspace="ws1")
> ++    store.add_role_permission(role.id, "workspace", "*", "READ")
> ++    store.assign_role_to_user(user.id, role.id)
> ++
> ++    # READ at the workspace level grants READ on every resource type in the workspace.
> ++    assert store.get_role_permission_for_resource(user.id, "experiment", "1", "ws1") == READ
> ++    assert store.get_role_permission_for_resource(user.id, "registered_model", "m1", "ws1") == READ
> ++    assert store.get_role_permission_for_resource(user.id, "gateway_endpoint", "e1", "ws1") == READ
> ++
> ++
> ++def test_workspace_permission_respects_union_with_specific(store, user):
> ++    role = store.create_role(name="mixed", workspace="ws1")
> ++    store.add_role_permission(role.id, "workspace", "*", "READ")
> ++    store.add_role_permission(role.id, "experiment", "42", "EDIT")
> ++    store.assign_role_to_user(user.id, role.id)
> ++
> ++    # Experiment 42: max(workspace READ, specific EDIT) = EDIT
> ++    assert store.get_role_permission_for_resource(user.id, "experiment", "42", "ws1") == EDIT
> ++    # Other experiments: just workspace READ
> ++    assert store.get_role_permission_for_resource(user.id, "experiment", "99", "ws1") == READ
> ++
> ++
> ++def test_is_workspace_admin(store, user):
> ++    role = store.create_role(name="ws-admin", workspace="ws1")
> ++    store.add_role_permission(role.id, "workspace", "*", "MANAGE")
> ++    store.assign_role_to_user(user.id, role.id)
> ++
> ++    assert store.is_workspace_admin(user.id, "ws1") is True
> ++    assert store.is_workspace_admin(user.id, "ws2") is False
> ++
> ++
> ++def test_is_workspace_admin_requires_manage(store, user):
> ++    # A non-MANAGE workspace permission does not make the user a WP admin.
> ++    role = store.create_role(name="ws-reader", workspace="ws1")
> ++    store.add_role_permission(role.id, "workspace", "*", "READ")
> ++    store.assign_role_to_user(user.id, role.id)
> ++
> ++    assert store.is_workspace_admin(user.id, "ws1") is False
> ++
>  +
>  +def test_get_role_permission_does_not_cross_workspace(store, user):
>  +    role = store.create_role(name="viewer", workspace="ws1")
> ````
>
> </details>


<sub>Reproduce locally: `git range-diff 1dfd617..21c0428 1dfd617..225bd25` | Disable: `git config gitstack.push-range-diff false`</sub>
</details>

**Comment by PattaraS** [comment]
/review

---

✅ [Review](https://github.com/mlflow/mlflow/actions/runs/24767068605?pr=22721) completed
<details><summary>Review Output</summary>

PR #22721 has been approved. Here's a summary of the review:

**Files reviewed**: 7 files (migration, models, entities, permissions, store, 2 test files)

**Key observations**:
- **Migration** correctly chains from `2ed73881770d` and creates `roles`, `role_permissions`, `user_role_assignments` with proper indexes, FKs, and unique constraints. Downgrade reverses cleanly.
- **SQLAlchemy models** use appropriate `cascade="all, delete-orphan"` on relationships, ensuring role deletion cleans up permissions and assignments.
- **Permission resolution** (`get_role_permission_for_resource`) correctly handles workspace-wide permissions, resource-specific permissions, wildcards, and union semantics using `selectinload` to avoid N+1 queries.
- **`is_workspace_admin`** is cleanly modeled as a permission triple check rather than a separate column — good design.
- **Tests** (48 cases) thoroughly cover CRUD, validation, cascades, workspace isolation, and permission resolution edge cases.

**No issues found** — code follows existing codebase patterns and style conventions.

</details>
